ENH: signal: add array API support / delegation to lfilter et al

ev-br opened the pull request:
<s>Is on top of https://github.com/scipy/scipy/pull/20772</s>

Add array API / delegation support to more routines from `signaltools`: lfilter, sosfilt, filtfilt, sosfiltfilt. Also `lfiltic` and `lfilt_zi`. While at it, also add `deconvolve` since it's only a small thing on top of `lfilter` anyway.

ev-br committed:
ENH: signal: array API support / delegation in lfilter-like functions

Note: Some CuPy functions deliberately deviate from scipy: {lfilter,sosfilt}_zi return
arrays of different shapes.

Skip dispatching for these cases; a user will have to be explicit on whether they want
the 'original' SciPy output (then call lfilter_zi with NumPy array arguments) or the
CuPy output (then call cupyx.scipy.signal.lfilter_zi with CuPy array arguments).

tylerjereddy commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3521,194 +3921,227 @@ def test_initial_conditions_3d_axis1(self, dt):
 
-        y, zf = sosfilt(sos, np.ones(40, dt), zi=zi)
-        assert_allclose_cast(zf, zi, rtol=1e-13)
+        y, zf = sosfilt(sos, xp.ones(40, dtype=dt), zi=zi)
```
Something interesting is happening with the test failure here via i.e., `python dev.py test -t scipy/signal/tests/test_signaltools.py::TestSOSFilt::test_sosfilt_zi -b cupy`

I noticed that `zf` in CuPy is the transpose of the result provided by NumPy and `torch`; i.e., this causes the above incantation to pass:

```diff
--- a/scipy/signal/tests/test_signaltools.py
+++ b/scipy/signal/tests/test_signaltools.py
@@ -3954,6 +3954,7 @@ class TestSOSFilt:
         with pytest.raises(ValueError, match='Invalid zi shape'):
             sosfilt(sos, x, zi=zi, axis=1)
 
+    @skip_xp_backends(cpu_only=True, exceptions=['cupy'])
     def test_sosfilt_zi(self, dt, xp):
         dt = getattr(xp, dt)
         sos = signal.butter(6, 0.2, output='sos')
@@ -3961,7 +3962,7 @@ class TestSOSFilt:
         zi = sosfilt_zi(sos)
 
         y, zf = sosfilt(sos, xp.ones(40, dtype=dt), zi=zi)
-        xp_assert_close(zf, zi, rtol=1e-13, check_dtype=False)
+        xp_assert_close(zf.T, zi, rtol=1e-13, check_dtype=False)
 
         # Expected steady state value of the step response of this filter:
         ss = xp.prod(xp.sum(sos[:, :3], axis=-1) / xp.sum(sos[:, 3:], axis=-1))
```

We obviously can't/shouldn't use a patch like that, filtered to CuPy case. At first glance, I can't imagine this would be an intentional change on the CuPy side based on the docs. I do see that the `zi` shape convention is a bit different between SciPy and CuPy, but upon closer inspection I wasn't entirely convinced that was the cause since `sosfilt_zi` should compensate by preserving the larger `zi` convention in CuPy anyway.

tylerjereddy reviewed (commented):
I noticed 213 new failures on GPU testing vs. `main` via `SCIPY_DEVICE=cuda python dev.py test -j 8 -b all`.

Many can be circumvented by skipping GPU tests with `torch` as usual, but I commented on a sample of one of the more interesting cases.

ev-br commented:
Hm. Apparently, `cupyx.scipy.signal.sosfilt_zi` is not a drop-in replacement for `scipy.signal.sosfilt_zi`:

1.  `sosfilt_zi` returns `zi` of different shapes in scipy and cupyx: in SciPy it's `(n_sections, 2)` and in `cupyx.scipy` it's `(n_sections, 4)`. Both are documented but the docs do not explain what these 2 and 4 are.
2. given identical `sos` inputs, the outputs of `scipy.signal.sosfilt_zi` and `cupyx.scipy.signal.sosfilt_zi` are numerically very different:

scipy:
```
(Pdb) p zi
array([[ 0.00525033, -0.00120091],
       [ 0.07729171, -0.02862319],
       [ 0.91711742, -0.65303262]])
(Pdb) a    
sos = array([[ 3.40537653e-04,  6.81075305e-04,  3.40537653e-04,
         1.00000000e+00, -1.03206941e+00,  2.75707942e-01],
       [ 1.00000000e+00,  2.00000000e+00,  1.00000000e+00,
         1.00000000e+00, -1.14298050e+00,  4.12801598e-01],
       [ 1.00000000e+00,  2.00000000e+00,  1.00000000e+00,
         1.00000000e+00, -1.40438489e+00,  7.35915191e-01]])
```

and `cupyx.scipy`:

```
(Pdb) p zi
array([[1.        , 1.        , 0.00559087, 0.00559087],
       [0.00559087, 0.00559087, 0.08288258, 0.08288258],
       [0.08288258, 0.08288258, 1.        , 1.        ]])
(Pdb) a
sos = array([[ 3.40537653e-04,  6.81075305e-04,  3.40537653e-04,
         1.00000000e+00, -1.03206941e+00,  2.75707942e-01],
       [ 1.00000000e+00,  2.00000000e+00,  1.00000000e+00,
         1.00000000e+00, -1.14298050e+00,  4.12801598e-01],
       [ 1.00000000e+00,  2.00000000e+00,  1.00000000e+00,
         1.00000000e+00, -1.40438489e+00,  7.35915191e-01]])
```

Before opening a cupy issue, let me ping @andfoy : as the author of cupy's `sosfilt` and `sosfilt_zi`, could you weigh in on the differences between scipy's and cupy's implementations ---  is this something to fix on the cupy side maybe?

andfoy commented:
The difference is that the state is not being represented as a sum of delays:

ev-br commented:
Thanks Edgar! This should probably be great to document, for both scipy and cupy.

That said, this does not explain the need to transpose noted in https://github.com/scipy/scipy/pull/21713#discussion_r1802139152 though?

andfoy commented:
I've just opened a PR in CuPy that fixes the issue at hand: https://github.com/cupy/cupy/pull/8677

ev-br commented:
Added  a commit with a slew of skips for CuPy, and a separate commit to avoid delegating for functions which deliberately deviate in scipy and cupy. 
I believe skipping tests is the right thing to do today; when Edgar's CuPy PR lands, we will be able to undo most of the skips, hopefully.

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2387,19 +2420,21 @@ def deconvolve(signal, divisor):
-    den = np.atleast_1d(divisor)
+    xp = array_namespace(signal, divisor)
+
+    num = xpx.atleast_nd(signal, ndim=1, xp=xp)
+    den = xpx.atleast_nd(divisor, ndim=1, xp=xp)
```
Nice that smoke-docs caught this! It demonstrates the need for additional array-like tests, though.
```suggestion
    num = xpx.atleast_nd(xp.asarray(signal), ndim=1, xp=xp)
    den = xpx.atleast_nd(xp.asarray(divisor), ndim=1, xp=xp)
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -4156,35 +4191,37 @@ def lfilter_zi(b, a):
     # b to be 2D.
-    b = np.atleast_1d(b)
+    b = xpx.atleast_nd(b, ndim=1, xp=xp)
     if b.ndim != 1:
         raise ValueError("Numerator b must be 1-D.")
-    a = np.atleast_1d(a)
+    a = xpx.atleast_nd(a, ndim=1, xp=xp)
```
need to handle array-likes here too

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -4156,35 +4191,37 @@ def lfilter_zi(b, a):
+        b = xp.concat((b, xp.zeros(n - b.shape[0], dtype=b.dtype)))
+
+    dt = xp.result_type(a, b)
+    IminusA = np.eye(n - 1) - linalg.companion(a).T
```
does this work, passing `a` to `linalg.companion` for non-NumPy arrays?

ev-br replied:
I suppose this goes through `__array__`. As long as it works, I'd keep it as is, until at some point `scipy.linalg.companion` gets array api compatible, and we'll fix this in one go.

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -4156,35 +4191,37 @@ def lfilter_zi(b, a):
     B = b[1:] - a[1:] * b[0]
     # Solve zi = A*zi + B
-    zi = np.linalg.solve(IminusA, B)
+    zi = xp.linalg.solve(IminusA, B)
```
```suggestion
    if hasattr(xp, linalg):
        zi = xp.linalg.solve(IminusA, B)
    else:
        zi = linalg.solve(np.asarray(IminusA), np.asarray(B))
        zi = xp.asarray(zi)        
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -4769,10 +4810,17 @@ def sosfilt(sos, x, axis=-1, zi=None):
+    if is_numpy(xp):
+        _reject_objects(sos, 'sosfilt')
+        _reject_objects(x, 'sosfilt')
+        if zi is not None:
+            _reject_objects(zi, 'sosfilt')
```
minor: could build `None` handling into `_reject_objects`?

ev-br committed:
BUG: signal: restore (and test for) array-like list inputs

The replacement for `np.atleast_1d(x)` is
`xpx.atleast_nd(xp.asarray(x), ndim=1, xp=xp)`, no less.

ev-br committed:
MAINT: signal: link to a cupy fix

ev-br commented:
Updated to address review comments, add some array-like tests, improve a link to a CuPy fix to investigate after a CuPy makes a 14.0 release, and added enough skips for tests to pass locally with SCIPY_DEVICE=cuda.

lucascolley commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2626,79 +2686,102 @@ def test_sine(self, xp):
-        xp_assert_equal(y0, np.swapaxes(y2, 0, 2))
+        y0 = self.filtfilt(zpk, x, padlen=0, axis=0, xp=xp)
+        y1 = self.filtfilt(
+            zpk, xp.asarray(np.swapaxes(x, 0, 1)), padlen=0, axis=1, xp=xp
```
Would this be equivalent?
```suggestion
            zpk, xp.moveaxis(x, [0, 1], [1, 0]), padlen=0, axis=1, xp=xp
```

ev-br replied:
I probably would. I'd still keep the test as is, for clarity. This is one of (rare) cases where swapaxes is more readable than moveaxis IMO.

lucascolley commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3912,117 +3991,165 @@ def test_nonnumeric_dtypes(func, xp):
 
         # Test simple FIR
-        b = np.array([1, 1]).astype(dt)
+        b = xp.asarray([1, 1],  dtype=dt)
```
```suggestion
        b = xp.asarray([1, 1], dtype=dt)
```

lucascolley commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3912,117 +3991,165 @@ def test_nonnumeric_dtypes(func, xp):
+        y_r = xp.asarray([0, 2, 4, 6, 8, 10.], dtype=dt)
         sos = tf2sos(b, a)
-        assert_array_almost_equal(sosfilt(tf2sos(b, a), x), y_r)
+        sos = xp.asarray(sos)   # XXX until tf2sos is numpy only
```
did you mean this? If so, the same appears below
```suggestion
        sos = xp.asarray(sos)   # XXX while tf2sos is numpy only
```

ev-br replied:
Oh, thanks! Apparently I changed the wording mid-phrase. Fixed in the last commit.

lucascolley commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -3912,117 +3991,165 @@ def test_nonnumeric_dtypes(func, xp):
 
-@skip_xp_backends(np_only=True)
-@pytest.mark.parametrize('dt', 'fdFD')
+@skip_xp_backends(cpu_only=True)
```
is it deliberate to be skipping these tests for CuPy too? (apologies if I haven't followed the conversation above)

ev-br replied:
It's deliberate, yes. There a collection of issues: a deliberate incompatibility of `_zi` functions, per Edgar's explanation above, and a collection of cupy bugs to be hopefully fixed when cupy 14 is released. The last commit adds a lengthy code comment.

lucascolley reviewed (commented):
looks pretty good, thanks!

ev-br committed:
address review comments

lucascolley reviewed (approved):
thanks Evgeni!
